src/Database/LSMTree/Internal/BloomFilterQuery2.hs

```diff
@@ -0,0 +1,441 @@
+-- inner loop, which causes all registered to be spilled to and restored from
+--the stack.
+--
+
+-- | A 'RunIxKeyIx' is a (compact) pair of a 'RunIx' and a 'KeyIx'.
+--
+-- We represent it as a 32bit word, using:
+--
+-- * 16 bits for the run\/filter index (MSB)
+-- * 16 bits for the key index (LSB)
+--
+newtype RunIxKeyIx = MkRunIxKeyIx Word32
+  deriving stock Eq
+  deriving newtype P.Prim
+
+pattern RunIxKeyIx :: RunIx -> KeyIx -> RunIxKeyIx
+pattern RunIxKeyIx r k <- (unpackRunIxKeyIx -> (r, k))
+  where
+    RunIxKeyIx r k = packRunIxKeyIx r k
+{-# INLINE RunIxKeyIx #-}
+{-# COMPLETE RunIxKeyIx #-}
+
+packRunIxKeyIx :: Int -> Int -> RunIxKeyIx
+packRunIxKeyIx r k =
+    assert (r >= 0 && r <= 0xffff
+         && k >= 0 && k <= 0xffff) $
+    MkRunIxKeyIx $
+      (fromIntegral :: Word -> Word32) $
+        (fromIntegral r `unsafeShiftL` 16)
+     .|. fromIntegral k
+{-# INLINE packRunIxKeyIx #-}
+
+unpackRunIxKeyIx :: RunIxKeyIx -> (Int, Int)
+unpackRunIxKeyIx (MkRunIxKeyIx c) =
+    ( fromIntegral (c `unsafeShiftR` 16)
+    , fromIntegral (c .&. 0xfff)
+    )
+{-# INLINE unpackRunIxKeyIx #-}
+
+instance Show RunIxKeyIx where
+  showsPrec _ (RunIxKeyIx r k) =
+    showString "RunIxKeyIx " . showsPrec 11 r
+              . showChar ' ' . showsPrec 11 k
+
+type Candidate = RunIxKeyIx
+
+-- | A candidate probe point is the combination of a filter\/run index,
+-- key index and hash number. This combination determines the probe location
+-- in the bloom filter, which is also cached.
+--
+-- We store these in 'PrimArray's as a pair of 64bit words:
+--
+-- * Low 64bit word:
+--   -    padding and the hash number (MSB)
+--   - 16 bits for the filter index
+--   - 16 bits for the key index      (LSB)
+--
+-- * High 64bit word: FilterBitIx
+--
+data CandidateProbe = MkCandidateProbe !Word64 !Word64
+
+type HashNo = Int
+type FilterBitIx = Int
+
+instance P.Prim CandidateProbe where
+    sizeOfType# _ = 16#
+    alignmentOfType# _ = 8#
+
+    indexByteArray# ba i =
+      MkCandidateProbe
+        (P.indexByteArray# ba (indexLo i))
+        (P.indexByteArray# ba (indexHi i))
+    readByteArray# ba i s1 =
+        case P.readByteArray# ba (indexLo i) s1 of { (# s2, lo #) ->
+        case P.readByteArray# ba (indexHi i) s2 of { (# s3, hi #) ->
+        (# s3, MkCandidateProbe lo hi #)
+        }}
+    writeByteArray# ba i (MkCandidateProbe lo hi) s =
+        P.writeByteArray# ba (indexHi i) hi
+       (P.writeByteArray# ba (indexLo i) lo s)
+
+    indexOffAddr# ba i =
+      MkCandidateProbe
+        (P.indexOffAddr# ba (indexLo i))
+        (P.indexOffAddr# ba (indexHi i))
+    readOffAddr# ba i s1 =
+        case P.readOffAddr# ba (indexLo i) s1 of { (# s2, lo #) ->
+        case P.readOffAddr# ba (indexHi i) s2 of { (# s3, hi #) ->
+        (# s3, MkCandidateProbe lo hi #)
+        }}
+    writeOffAddr# ba i (MkCandidateProbe lo hi) s =
+        P.writeOffAddr# ba (indexHi i) hi
+       (P.writeOffAddr# ba (indexLo i) lo s)
```

Comment:
Oh nice. I didn't know about that.

Reply:
We've caught a bug before in a `Prim` instance that we defined `lsm-tree` by checking `primLaws`. Often I also add a property test that runs the laws, of which we have an example in the repo: 

https://github.com/IntersectMBO/lsm-tree/blob/9d3d730b6171b34ad12cd90faf3fc6a6b9b6843c/test/Test/Database/LSMTree/Internal/Entry.hs#L30-L37

It just needs a bit of boilerplate, since `quickcheck-classes` exposes `IO` runners instead of `Property` runners

Reply:
Nice. Done.